torch/csrc/autograd/python_variable_indexing.cpp

```diff
@@ -181,79 +146,49 @@ static Variable applySlicing(const Variable& self, PyObject* index, variable_lis
   for (int64_t i = 0; i < size; i++) {
     PyObject* obj = PyTuple_GET_ITEM(index, i); // NOLINT(cppcoreguidelines-pro-type-cstyle-cast)
     if (THPUtils_checkLong(obj)) {
-      result = applySelect(result, dim, obj, i);
+      if (THPVariable_Check(obj)) {
+        recordSelectTrace(THPVariable_Unpack(obj));
+      }
+      result = at::indexing::handleInteger(result, dim, THPUtils_unpackLong(obj), i);
     } else if (PySlice_Check(obj)) {
-      result = applySlice(result, dim, obj);
-      dim++;
+      Py_ssize_t start, stop, step;
+      Tensor start_tensor, stop_tensor, step_tensor;
+      unpackSliceAndExtractTensors(obj, start, stop, step, start_tensor, stop_tensor, step_tensor);
+      recordSliceTrace(start_tensor, stop_tensor, step_tensor);
+      result = at::indexing::handleSlice(result, dim, start, stop, step, torch::jit::tracer::isTracing());
     } else if (obj == Py_Ellipsis) {
-      dim += self.dim() - specified_dims;
+      at::indexing::handleEllipsis(self, dim, specified_dims);
     } else if (obj == Py_None) {
-      result = result.unsqueeze(dim);
-      dim++;
+      result = at::indexing::handleNone(result, dim);
     } else if (PyBool_Check(obj)) {
-      result = result.unsqueeze(dim);
-      handle_var(boolToIndexingTensor(result, obj == Py_True)); // NOLINT(cppcoreguidelines-pro-type-cstyle-cast)
+      result = at::indexing::handleBoolean(result, obj == Py_True, outIndices, dim);
     } else if (THPVariable_Check(obj)) {
-      auto& var = THPVariable_Unpack(obj);
-      auto scalar_type = var.scalar_type();
-      if (var.dim() == 0 && at::isIntegralType(scalar_type, /*includeBool=*/true)) {
-        if (scalar_type != at::kByte && scalar_type != at::kBool) {
-          result = applySelect(result, dim, obj, i);
-        } else {
-          result = result.unsqueeze(dim);
-          if(scalar_type == at::kBool) {
-            handle_var(boolToIndexingTensor(result, var.item<bool>() != 0));
-          } else {
-            handle_var(boolToIndexingTensor(result, var.item<uint8_t>() != 0));
-          }
-        }
-      } else {
-        handle_var(var);
+      Tensor tensor = THPVariable_Unpack(obj);
+      auto scalar_type = tensor.scalar_type();
+      if (tensor.dim() == 0 && at::isIntegralType(scalar_type, /*includeBool=*/false) && scalar_type != at::kByte) {
+        recordSelectTrace(tensor);
       }
+      result = at::indexing::handleTensor(result, tensor, outIndices, dim, i);
```

Comment:
I guess, this is one more payoff from factoring out this logic into a helper function?